### ISSUE TITLE
Models for subdocuments. 

### ISSUE THREAD

**Issue by amit777** [vdemedes/mongorito#43]
This project looks like it has almost everything I need except for th ability to have sub document models. Or am I missing something?

**Comment by vdemedes** [comment]
There's no ability to save sub-models, but you can save document with nested fields. Is that what you want?

**Comment by amit777** [comment]
Well let's say I have a domain model. And within domain model I have an array of server documents which I want to have a model for.  Very much like mongoose child schema models.

**Comment by vdemedes** [comment]
No, this feature is not supported currently.

**Comment by amit777** [comment]
I actually just realized the mongoose does not allowed for subjdocument child schema models.  They do allow subdocument arrays of models though:
eg:

Not Allowed:
```
parentSchema = new Schema({
attr: 'hi',
child:  childSchema  // not allowed
children: [childSchema] //allowed
})
```

**Comment by jaredpetersen** [comment]
:+1: -- I was thinking of switching over from Mongoose to Mongorito but I can't without this feature.

**Comment by Alex-iFactory** [comment]
Is this possible yet?

**Comment by vdemedes** [comment]
No, not yet. If this functionality is critical for you, I recommend going with Mongoose, because I'm very short on time lately, so I can't promise any ETA.

**Comment by Alex-iFactory** [comment]
@vdemedes I guess I can do a pull request with some of the code I have come up with.

I have one question you might be able to help with, how would you go about figuring out where the sub document class is. With `Doc.get('sub').get('test')` you can load the entity into the parent class because it is already imported; But how do we figure out what class to load the sub into?

Cheers

**Comment by Alex-iFactory** [comment]
Actually i guess it could be done with `module.parent` however not the best solution... It would require modules never change their directory

**Comment by Alex-iFactory** [comment]
Something like:

https://github.com/Alex-iFactory/eden/blob/master/bin/bundles/core/model.js

May work, not sure how to make the getAttributes more efficient?

**Comment by vdemedes** [comment]
Not sure I understand your question, could you clarify?

**Comment by Alex-iFactory** [comment]
If we were to allow sub-documents; wouldn't when we run a query on a parent document we need to require the sub-document class:

```jsx
class parentDocument extends mongorito.Model {

}

class childDocument extends mongorito.Model {

}

let ParentDocument = new parentDocument({
    'child' : yield new childDocument({
        'title' : 'test'
    }).save();
});
```

In the above case, finding the parent document would return _information_ about the child document, but I am unsure how to use this information to require and load the `childDocument` class. What would you use to store the location of the `childDocument` class in the `parentDocument` document?

Cheers

**Comment by vdemedes** [comment]
Oh I see now. We could modify `Query#populate()` to convert inline data to models straight away. Right now `populate()` works just like in Mongoose. Given this document:

```js
{
  "comments": [1, 2, 3]
}
```

where `[1, 2, 3]` are ObjectIDs, we can find documents with those ids and create a `Comment` model for each of them:

```js
let post = await Post.populate('comments', Comment).findOne();
// post.comments now is an array of Comment models
```

So we could modify implementation of `.populate()` to avoid `find` operation and create a new model straight away, if current array item is not an ObjectID:

```js
if (isObjectId(item)) {
  return model.findById(item);
} else {
  return new model(item);
}
```

Let me know if that's unclear ;)

**Comment by Alex-iFactory** [comment]
That works, I will put together a pull request this week.

Cheers

**Comment by vdemedes** [comment]
Cool!

**Comment by sascha1337** [comment]
is this merged yet?

**Comment by vdemedes** [comment]
Nope, it was never implemented. Hope to implement this in the next major version - #138.
